units/historical_test_data.go

```diff
@@ -0,0 +1,206 @@
+package units
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/evergreen-ci/cedar"
+	"github.com/evergreen-ci/cedar/cost"
+	"github.com/evergreen-ci/cedar/model"
+	"github.com/evergreen-ci/utility"
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/amboy/dependency"
+	"github.com/mongodb/amboy/job"
+	"github.com/pkg/errors"
+)
+
+const historicalTestDataJobName = "historical-test-data"
+
+type historicalTestDataJob struct {
+	job.Base `bson:"job_base" json:"job_base" yaml:"job_base"`
+	Info     model.HistoricalTestDataInfo `bson:"info" json:"info" yaml:"info"`
+	Result   model.TestResult             `bson:"result" json:"result" yaml:"result"`
+
+	env      cedar.Environment
+	settings historicalTestDataProjectSettings
+}
+
+// NewHistoricalTestDataJob returns a job that re-computes and stores the
+// historical test data based on the new incoming test result.
+func NewHistoricalTestDataJob(env cedar.Environment, info model.TestResultsInfo, tr model.TestResult) amboy.Job {
+	j := makeHistoricalTestDataJob()
+
+	j.env = env
+	j.Result = tr
+	taskName := j.getTaskName(info)
+	j.Info = model.HistoricalTestDataInfo{
+		Project:     info.Project,
+		Variant:     info.Variant,
+		TaskName:    taskName,
+		TestName:    tr.TestName,
+		RequestType: info.RequestType,
+		Date:        tr.TestEndTime.UTC(),
+	}
+
+	j.SetScopes([]string{
+		j.Info.Project,
+		j.Info.Variant,
+		taskName,
+		j.Info.TestName,
+		j.Info.RequestType,
+		j.Info.Date.Format(model.HistoricalTestDataDateFormat),
+	})
+
+	return j
+}
+func makeHistoricalTestDataJob() *historicalTestDataJob {
+	j := &historicalTestDataJob{
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    historicalTestDataJobName,
+				Version: 0,
+			},
+		},
+	}
+	j.SetDependency(dependency.NewAlways())
+	return j
+}
+
+// getTaskName returns the task name to attribute to the. The display task name
+// is always prioritized over the execution task name, if the task is part of a
+// display task.
+func (j *historicalTestDataJob) getTaskName(info model.TestResultsInfo) string {
+	if info.DisplayTaskName != "" {
+		return info.DisplayTaskName
+	}
+	return info.TaskName
+}
+
+func (j *historicalTestDataJob) Run(ctx context.Context) {
+	defer j.MarkComplete()
+
+	if j.env == nil {
+		j.env = cedar.GetEnvironment()
+	}
+
+	conf := model.NewCedarConfig(j.env)
+	if err := conf.Find(); err != nil {
+		j.AddError(errors.Wrap(err, "finding cedar configuration"))
+		return
+	}
+	if conf.Flags.DisableHistoricalTestData {
+		return
+	}
+
+	skip, err := j.shouldNoop(ctx, conf)
+	j.AddError(errors.Wrapf(err, "checking if job needs to run"))
```

Comment:
We want to continue on error here?

Reply:
Yeah, `skip` is what actually determines if the job should continue.